### ISSUE TITLE
Triggering cache refresh repeatedly causes duplicated refresh jobs

### ISSUE THREAD

**Issue by paweltatarczuk** [putyourlightson/craft-blitz#245]
### Describe the bug
When elements are being saved in a loop and for each element the cache is being invalidated duplicated refresh jobs are created:

- Job 1: refresh element 1
- Job 2: refresh element 1, 2
- Job 3: refresh element 1, 2, 3
...

### To reproduce
Steps to reproduce the behaviour:
1. Add element to RefreshCacheService
2. Run refresh() on RefreshCacheService twice

For example:

```
$entry = Entry::findOne();
Blitz::$plugin->refreshCache->addElement($entry);

Blitz::$plugin->refreshCache->refresh();
Blitz::$plugin->refreshCache->refresh();
```

### Expected behaviour
Executing `RefreshCacheService::refresh()` multiple times should not create duplicated refresh jobs.

### Screenshots
Here's the log that shows what is happening:

```
2020-08-18 15:00:24 [][][info] {"cacheIds":["8852","8850"],"elements":{"craft\\commerce\\elements\\Variant":{"elementIds":[],"sourceIds":[]},"craft\\commerce\\elements\\Product":{"elementIds":[192447],"sourceIds":["8"]}},"clearCache":true,"description":null}
2020-08-18 15:00:24 [][][info] {"cacheIds":["8852","8850"],"elements":{"craft\\commerce\\elements\\Variant":{"elementIds":[],"sourceIds":[]},"craft\\commerce\\elements\\Product":{"elementIds":[192447,535084],"sourceIds":["8"]}},"clearCache":true,"description":null}
2020-08-18 15:00:24 [][][info] {"cacheIds":["8852","8850"],"elements":{"craft\\commerce\\elements\\Variant":{"elementIds":[],"sourceIds":[]},"craft\\commerce\\elements\\Product":{"elementIds":[192447,535084,160114],"sourceIds":["8"]}},"clearCache":true,"description":null}
2020-08-18 15:00:25 [][][info] {"cacheIds":["8852","8850"],"elements":{"craft\\commerce\\elements\\Variant":{"elementIds":[],"sourceIds":[]},"craft\\commerce\\elements\\Product":{"elementIds":[192447,535084,160114,160104],"sourceIds":["8"]}},"clearCache":true,"description":null}
...
```

Excerpt of the stack trace:

```
#0 /app/vendor/yiisoft/yii2/base/BaseObject.php(109): putyourlightson\blitz\jobs\RefreshCacheJob->init()
#1 /app/vendor/putyourlightson/craft-blitz/src/services/RefreshCacheService.php(390): yii\base\BaseObject->__construct(Array)
#2 /app/vendor/putyourlightson/craft-blitz/src/drivers/integrations/SeomaticIntegration.php(64): putyourlightson\blitz\services\RefreshCacheService->refresh()
#3 [internal function]: putyourlightson\blitz\drivers\integrations\SeomaticIntegration::putyourlightson\blitz\drivers\integrations\{closure}(Object(nystudio107\seomatic\events\InvalidateContainerCachesEvent))
#4 /app/vendor/yiisoft/yii2/base/Event.php(312): call_user_func(Object(Closure), Object(nystudio107\seomatic\events\InvalidateContainerCachesEvent))
#5 /app/vendor/yiisoft/yii2/base/Component.php(636): yii\base\Event::trigger('nystudio107\\seo...', 'invalidateConta...', Object(nystudio107\seomatic\events\InvalidateContainerCachesEvent))
#6 /app/vendor/nystudio107/craft-seomatic/src/services/MetaContainers.php(756): yii\base\Component->trigger('invalidateConta...', Object(nystudio107\seomatic\events\InvalidateContainerCachesEvent))
#7 /app/vendor/nystudio107/craft-seomatic/src/services/MetaBundles.php(425): nystudio107\seomatic\services\MetaContainers->invalidateContainerCacheByPath('', '3')
#8 /app/vendor/nystudio107/craft-seomatic/src/Seomatic.php(556): nystudio107\seomatic\services\MetaBundles->invalidateMetaBundleByElement(Object(craft\commerce\elements\Product), false)
#9 [internal function]: nystudio107\seomatic\Seomatic->nystudio107\seomatic\{closure}(Object(craft\events\ElementEvent))
#10 /app/vendor/yiisoft/yii2/base/Event.php(312): call_user_func(Object(Closure), Object(craft\events\ElementEvent))
#11 /app/vendor/yiisoft/yii2/base/Component.php(636): yii\base\Event::trigger('craft\\services\\...', 'afterSaveElemen...', Object(craft\events\ElementEvent))
#12 /app/vendor/craftcms/cms/src/services/Elements.php(2261): yii\base\Component->trigger('afterSaveElemen...', Object(craft\events\ElementEvent))
#13 /app/vendor/craftcms/cms/src/services/Elements.php(2380): craft\services\Elements->_saveElementInternal(Object(craft\commerce\elements\Product), true, false)
#14 /app/vendor/craftcms/cms/src/services/Elements.php(2178): craft\services\Elements->_propagateElement(Object(craft\commerce\elements\Product), Array, Object(craft\commerce\elements\Product))
#15 /app/vendor/craftcms/cms/src/services/Elements.php(531): craft\services\Elements->_saveElementInternal(Object(craft\commerce\elements\Product), true, true, true)
```

### Versions
- Plugin version: 3.6.5
- Craft version: 3.4.27

**Comment by bencroker** [comment]
What you are experiencing is the expected behaviour. You should enable `batchMode` to avoid multiple refresh jobs being created in the same request.

```php
Blitz::$plugin->refreshCache->batchMode = true;

Blitz::$plugin->refreshCache->addElement($entry1);
Blitz::$plugin->refreshCache->addElement($entry2);
Blitz::$plugin->refreshCache->addElement($entry3);

Blitz::$plugin->refreshCache->refresh();
```

**Comment by paweltatarczuk** [comment]
@bencroker Let me rephrase the issue:

Let's say we update multiple elements in a loop:

```
$elements = [ ... ]
foreach ($elements as $element) {
  Craft::$app->elements->saveElement($element);
}
```

For each saved element a RefreshCacheJob is created, every following job is refreshing elements saved in a previous loop iterations.

If we save 1000 elements we get 1000 refresh jobs with 1, 2, 3, ... 1000 elements being refreshed.

After the changes in #246 this issue does not occur any more. I guess I should be able to save elements in a loop without creating duplicated refresh jobs.

**Comment by bencroker** [comment]
As stated, that is what `batchMode` is intended for. https://github.com/putyourlightson/craft-blitz/pull/246 may fix your issue, but it may also break other aspects of how Blitz works. 

```php
Blitz::$plugin->refreshCache->batchMode = true;

foreach ($elements as $element) {
  Craft::$app->elements->saveElement($element);
}

Blitz::$plugin->refreshCache->refresh();
```

**Comment by paweltatarczuk** [comment]
Does this mean I need to enable batchMode wherever I want to update elements in a loop? This would couple my module to craft-blitz module. I mean why should another module know about Blitz internals?

I would expect it to create 1k jobs to refresh cache for 1 element each and not 1k jobs where the last one refreshes 1k elements. Something is not right here.

**Comment by bencroker** [comment]
Ok I see now what you mean regarding the jobs incrementally getting larger. I'll take a closer look at the PR you submitted and see how it affects the rest of the system.

**Comment by bencroker** [comment]
Fixed in https://github.com/putyourlightson/craft-blitz/commit/44722e2a55a250a13736508bd726c1ec628bd76c for the next release, thanks!

**Comment by paweltatarczuk** [comment]
Thank you!

**Comment by bencroker** [comment]
Released in version 3.6.8.
